File: builder/builder.php
<?php

interface Builder
{
    public function toSql(): string;
    public function where(string $column, string $operator, string $value): self;
    public function select(...$selects): self;
    public static function table(string $table): self;
    public function orderBy(string $column, string $type = "ASC"): self;
    public function groupBy(...$columns): self;
    public function limit(int $limit, int $offset = 0): self;
    public function get(): array;
    public function first(): array;
}

class MysqlBuilder implements Builder
{

    public function toSql(): string 
    {

        return "";
    }
    public function where(string $column, string $operator, string $value): self
    {

        return $this;
    }
    public function select(...$selects): self 
    {


        return $this;

    }
    public static function table(string $table): self
    {

        return new self();
    }
    public function orderBy(string $column, string $type = "ASC"): self
    {

        return $this;
    }
    public function groupBy(...$columns): self
    {

        return $this;
    }

    public function get(): array
    {

        return [];
    }

    public function first(): array
    {

        return [];
    }

    public function limit(int $limit, int $offset = 0): self
    {

        return new self();
    }
}

// Client implementation
class Post {
    protected Builder $builder;

    public function __construct()
    {       
        $this->builder = new MysqlBuilder();
    }

    public  function getPosts(string $search): array {
        return $this->builder->table("posts")->select("id", "title", "content")
            ->where("title", "LIKE", $search)
            ->limit(10, 0)
            ->get();
    }
}

File: builder/builder.go
package main

type Builder interface {
	Table(string)
	Select(...string)
	Where(string, string, string)
	Limit(int, int)
	OrderBy(string, string)
}

type Mysqlbuilder struct {
	table     string
	selects   []string
	wheres    []string
	limit     string
	orderBy   string
	direction string
}

func (b *Mysqlbuilder) Table(table string) {
	b.table = table
}

func (b *Mysqlbuilder) Select(selects ...string) {
	b.selects = selects
}

func (b *Mysqlbuilder) Where(field string, operator string, value string) {
	b.wheres = append(b.wheres, field+" "+operator+" "+value)
}

func (b *Mysqlbuilder) Limit(offset int, limit int) {
	b.limit = string(offset) + "," + string(limit)
}

func (b *Mysqlbuilder) OrderBy(field string, direction string) {
	b.orderBy = field
	b.direction = direction
}

func (b *Mysqlbuilder) GetSQL() string {
	sql := "SELECT "
	if len(b.selects) > 0 {
		for i, s := range b.selects {
			sql += s
			if i < len(b.selects)-1 {
				sql += ", "
			}
		}
	} else {
		sql += "*"
	}

	sql += " FROM " + b.table

	if len(b.wheres) > 0 {
		sql += " WHERE "
		for i, w := range b.wheres {
			sql += w
			if i < len(b.wheres)-1 {
				sql += " AND "
			}
		}
	}

	if b.orderBy != "" {
		sql += " ORDER BY " + b.orderBy + " " + b.direction
	}

	if b.limit != "" {
		sql += " LIMIT " + b.limit
	}

	return sql
}

// Client impelementation

// This is very basic implementation you can implement director struct and call it on client method
// and it gives more simplicity

func main() {
	var builder Builder

	builder = &Mysqlbuilder{}
	builder.Table("users")
	builder.Select("id", "name", "email")
	builder.Where("id", "=", "1")
	builder.Limit(0, 10)
	builder.OrderBy("id", "asc")

}
